From a44e9fb60e478039b5f62cb27f3652a28014898d Mon Sep 17 00:00:00 2001
From: Rene Kita <mail@rkta.de>
Date: Sun, 26 Dec 2021 17:28:44 +0100
Subject: [PATCH 078/172] Remove all usage of variable length arrays

Variable length arrays are optional since C11 and can be problematic.
We are already using a garbage collector - use it instead of VLAs. This
also allows us to remove some #ifdef's.
---
 table.c |   35 ++++++++---------------------------
 1 file changed, 8 insertions(+), 27 deletions(-)

--- a/table.c
+++ b/table.c
@@ -1252,16 +1252,10 @@ check_table_width(struct table *t, doubl
     int i, j, k, m, bcol, ecol;
     int corr = 0;
     struct table_cell *cell = &t->cell;
-#ifdef __GNUC__
-    short orgwidth[t->maxcol >= 0 ? t->maxcol + 1 : 1];
-    short corwidth[t->maxcol >= 0 ? t->maxcol + 1 : 1];
-    short cwidth[cell->maxcell >= 0 ? cell->maxcell + 1 : 1];
-    double swidth[cell->maxcell >= 0 ? cell->maxcell + 1 : 1];
-#else				/* __GNUC__ */
-    short orgwidth[MAXCOL], corwidth[MAXCOL];
-    short cwidth[MAXCELL];
-    double swidth[MAXCELL];
-#endif				/* __GNUC__ */
+    short *orgwidth = GC_malloc(t->maxcol >= 0 ? t->maxcol + 1 : 1);
+    short *corwidth = GC_malloc(t->maxcol >= 0 ? t->maxcol + 1 : 1);
+    short *cwidth = GC_malloc(cell->maxcell >= 0 ? cell->maxcell + 1 : 1);
+    double *swidth = GC_malloc(cell->maxcell >= 0 ? cell->maxcell + 1 : 1);
     double twidth, sxy, *Sxx, stotal;
 
     twidth = 0.;
@@ -1606,11 +1600,7 @@ check_table_height(struct table *t)
 static int
 get_table_width(struct table *t, short *orgwidth, short *cellwidth, int flag)
 {
-#ifdef __GNUC__
-    short newwidth[t->maxcol >= 0 ? t->maxcol + 1 : 1];
-#else				/* not __GNUC__ */
-    short newwidth[MAXCOL];
-#endif				/* not __GNUC__ */
+    short *newwidth = GC_malloc(t->maxcol >= 0 ? t->maxcol + 1 : 1);
     int i;
     int swidth;
     struct table_cell *cell = &t->cell;
@@ -1620,11 +1610,7 @@ get_table_width(struct table *t, short *
 	newwidth[i] = max(orgwidth[i], 0);
 
     if (flag & CHECK_FIXED) {
-#ifdef __GNUC__
-	short ccellwidth[cell->maxcell >= 0 ? cell->maxcell + 1 : 1];
-#else				/* not __GNUC__ */
-	short ccellwidth[MAXCELL];
-#endif				/* not __GNUC__ */
+	short *ccellwidth = GC_malloc(cell->maxcell >= 0 ? cell->maxcell + 1 : 1);
 	for (i = 0; i <= t->maxcol; i++) {
 	    if (newwidth[i] < t->fixed_width[i])
 		newwidth[i] = t->fixed_width[i];
@@ -3496,13 +3482,8 @@ set_table_matrix0(struct table *t, int m
     int i, j, k, bcol, ecol;
     int width;
     double w0, w1, w, s, b;
-#ifdef __GNUC__
-    double we[size];
-    char expand[size];
-#else				/* not __GNUC__ */
-    double we[MAXCOL];
-    char expand[MAXCOL];
-#endif				/* not __GNUC__ */
+    double *we = GC_malloc(size);
+    char *expand = GC_malloc(size);
     struct table_cell *cell = &t->cell;
 
     w0 = 0.;
